### PR TITLE
Added version 4 signature calculations

### PR DESCRIPTION
**Pull request by monty16597**
Added Signature calculation algorithm for version 4 pre-sign URL.
Added test cases for the different types of methods and different types of requests (valid, invalid and expired).
#3522 
#3266

### CODE REVIEW

File: localstack/services/s3/s3_listener.py [hunk 1]
```diff
@@ -1494,5 +1552,45 @@ def authenticate_presign_url(method, path, headers, data=None):
         )
 
 
+def authenticate_presign_url_signv4(method, path, headers, data, url, query_params, request_dict):
+
+    # Calculating Signature
+    LOGGER.debug('Calculating the version 4 signature.')
```
Comment (whummer) [inline localstack/services/s3/s3_listener.py:1558]: I'd say let's remove this log output (to avoid being too verbose).

**Review by whummer — COMMENTED** [review]
Looks great, thanks @monty16597 ! 🚀  Only added a few minor comments. Thanks

File: localstack/services/s3/s3_listener.py [hunk 2]
```diff
@@ -1456,9 +1485,41 @@ def authenticate_presign_url(method, path, headers, data=None):
...
+                X-Amz-Credential, X-Amz-Date, X-Amz-Expires, \
+                X-Amz-SignedHeaders and X-Amz-Signature parameters.',
+            code_string='AccessDenied'
+        )
+    elif is_v4 and not is_v2:
+        LOGGER.debug('Valid version 4 presign URL.')
+        response = authenticate_presign_url_signv4(method, path, headers, data, url, query_params, request_dict)
+
+    if response is not None:
+        LOGGER.debug('Signature calculation failed with the error.')
```
[the remark this answers is not in the document]
Comment (monty16597) [reply localstack/services/s3/s3_listener.py:1514]: Actually, let's add the error log here. So people can have errors while testing presign URL. thoughts?

File: localstack/services/s3/s3_listener.py [hunk 3]
```diff
@@ -110,7 +111,21 @@
     'Access-Control-Request-Headers', 'Access-Control-Request-Method'
 ]
 
-SIGNATURE_V4_PARAMS = ['X-Amz-Algorithm']
+SIGNATURE_V4_PARAMS = [
+    'X-Amz-Algorithm', 'X-Amz-Credential', 'X-Amz-Date', 'X-Amz-Expires',
+    'X-Amz-SignedHeaders', 'X-Amz-Signature'
+]
+
+SIGNATURE_QUERY_STRING = [
```
[the remark this answers is not in the document]
Comment (monty16597) [reply localstack/services/s3/s3_listener.py:119]: Yes, you are right we have some changes in the logic. So we are not using those variables anymore inside the code. I think we are good to remove those variables.

**Review by whummer — COMMENTED** [review]
Looks great, thanks @monty16597 👍

**Pull request merged**